### PR TITLE
Launch DDS from Dart SDK and prepare to serve DevTools from DDS

### PR DESCRIPTION
**Pull request by bkonyi**
This change is a major step towards moving away from shipping DDS via Pub.

The first component of this PR is the move away from importing package:dds to launch DDS. Instead, DDS is launched out of process using the `dart development-service` command shipped with the Dart SDK. This makes Flutter's handling of DDS consistent with the standalone Dart VM.

The second component of this PR is the initial work to prepare for the removal of instances of DevTools being served manually by the flutter_tool, instead relying on DDS to serve DevTools. This will be consistent with how the standalone Dart VM serves DevTools, tying the DevTools lifecycle to a live DDS instance. This will allow for the removal of much of the logic needed to properly manage the lifecycle of the DevTools server in a future PR. Also, by serving DevTools from DDS, users will no longer need to forward a secondary port in remote workflows as DevTools will be available on the DDS port.

There's two remaining circumstances that will prevent us from removing DevtoolsRunner completely:

 - The daemon's `devtools.serve` endpoint
 - `flutter drive`'s `--profile-memory` flag used for recording memory profiles

This PR also includes some refactoring around `DebuggingOptions` to reduce the number of debugging related arguments being passed as parameters adjacent to a `DebuggingOptions` instance.

### CODE REVIEW

**Comment by bkonyi** [comment]
I think this is ready for a first pass. The failing integration test will be fixed once the Dart SDK roll picks up [this CL](https://dart-review.googlesource.com/c/sdk/+/362401). I still need to investigate what changes are needed for google3.

**Comment by christopherfujino** [comment]
I scheduled time next week Monday to take a look at this tome :)

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (kenzieschmoll) [inline packages/flutter_tools/lib/src/base/dds.dart:294]: dart doc for these members and methods?

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (kenzieschmoll) [inline packages/flutter_tools/lib/src/base/dds.dart:361]: nit: no need for trailing comma here?

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (kenzieschmoll) [inline packages/flutter_tools/lib/src/base/dds.dart:375]: nit: trailing comma?

File: packages/flutter_tools/lib/src/commands/test.dart
Comment (kenzieschmoll) [inline packages/flutter_tools/lib/src/commands/test.dart:410]: we seem to repeat a lot of this arg parsing logic throughout flutter tools. Could be nice if there were some common getters to use

```
bool get printDtdFlag => boolArg(FlutterGlobalOptions.kPrintDtd, global: true);
bool get enableDevToolsFlag => boolArg(FlutterCommand.kEnableDevTools);
bool get disableServiceAuthCodesFlag => boolArg('disable-service-auth-codes');
```

but likely out of scope for this PR as this would be a larger cleanup

File: packages/flutter_tools/lib/src/resident_runner.dart
Comment (kenzieschmoll) [inline packages/flutter_tools/lib/src/resident_runner.dart:1511]: can you add back the new line so that the DevTools URI is separated in the CLI output?

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (kenzieschmoll) [inline packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:179]: where is the `--print-dtd` behavior test covered by flutter tools now?

**Comment by christopherfujino** [comment]
@bkonyi As I understand it, this PR is doing two different things:

1. Changing the way the tool launches DDS from calling `dds.DartDevelopmentService.startDartDevelopmentService` from package:dds to spawning a sub-process via `dart development-service`
2. Changing the way the tool launches DevTools from directly to via DDS

Do I have this right? And if so, would it be possible to split this into two separate PRs, both to make it easier to review and to minimize the impact if either of these changes needs to be reverted?

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (bkonyi) [reply packages/flutter_tools/lib/src/base/dds.dart:71]: It's included by default if DTD is started. `--print-dtd` only impacts whether or not the standalone VM prints the URI once DDS starts.

File: packages/flutter_tools/lib/src/resident_runner.dart
Comment (bkonyi) [reply packages/flutter_tools/lib/src/resident_runner.dart:1511]: I removed it so the DevTools URI was grouped with the VM service and DTD messages. I can print a newline after all these messages are printed instead.

File: packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart
Comment (bkonyi) [reply packages/flutter_tools/test/general.shard/runner/flutter_command_runner_test.dart:179]: There wasn't an obvious way to test this after moving it out of a global variable and into `DebuggingOptions`. I've reverted the change for this PR and added back the test, but will investigate options for testing in a follow up PR when I remove the global flag.

**Comment by bkonyi** [comment]
> @bkonyi As I understand it, this PR is doing two different things:
> 
> 1. Changing the way the tool launches DDS from calling `dds.DartDevelopmentService.startDartDevelopmentService` from package:dds to spawning a sub-process via `dart development-service`
> 2. Changing the way the tool launches DevTools from directly to via DDS
> 
> Do I have this right? And if so, would it be possible to split this into two separate PRs, both to make it easier to review and to minimize the impact if either of these changes needs to be reverted?

I've gone ahead and reverted to using the `ResidentDevToolsHandler` code to manage the DevTools instance to reduce the surface area of this change. Much of the code needed to move over to serving DevTools from DDS is still present but commented out with TODOs and can be reviewed in the follow up PR.

There may be some tests still failing, but I think this is ready for a first pass.

**Comment by christopherfujino** [comment]
> > @bkonyi As I understand it, this PR is doing two different things:
> > 
> > 1. Changing the way the tool launches DDS from calling `dds.DartDevelopmentService.startDartDevelopmentService` from package:dds to spawning a sub-process via `dart development-service`
> > 2. Changing the way the tool launches DevTools from directly to via DDS
> > 
> > Do I have this right? And if so, would it be possible to split this into two separate PRs, both to make it easier to review and to minimize the impact if either of these changes needs to be reverted?
> 
> I've gone ahead and reverted to using the `ResidentDevToolsHandler` code to manage the DevTools instance to reduce the surface area of this change. Much of the code needed to move over to serving DevTools from DDS is still present but commented out with TODOs and can be reviewed in the follow up PR.
> 
> There may be some tests still failing, but I think this is ready for a first pass.

Thanks. I'm pretty swamped today, but scheduled time tomorrow to review.

**Comment by andrewkolos** [comment]
> No rush, the google3 changes required for this to land are blocked on Dart CLI support in google3, so I'm not sure how long that'll take to clear up.

@bkonyi is this still the case?

**Comment by bkonyi** [comment]
> > No rush, the google3 changes required for this to land are blocked on Dart CLI support in google3, so I'm not sure how long that'll take to clear up.
> 
> @bkonyi is this still the case?

Yeah, I've been fighting with some build rules to get all the right artifacts shipped with the Dart SDK packaged with the Flutter extracted and depended on properly. It's been a headache so far :(

**Comment by bkonyi** [comment]
FYI @christopherfujino @kenzieschmoll, this should be ready for review. This PR is currently blocked from landing as [this CL in the Dart SDK](https://dart-review.googlesource.com/c/sdk/+/370000) needs to be rolled into google3, but otherwise this change is good to go.

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/dds.dart:41]: Can you instead require the path to the dart executable as a function argument?

Then at the callsite of this function, you can get the path via something like:

```dart
final String dartPath = globals.artifacts.getArtifactPath(Artifact.engineDartBinary);
```

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/dds.dart:50]: do you have a tracking issue for this?

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/dds.dart:65]: I chatted with bquinlan, and we agreed that this is not safe--decoding JSON *may* fail if the first event emitted by the stderr stream does not include ALL the bytes of the JSON object.

If the assumption is that this is the only thing `dart development-service` will output to STDERR is this JSON blob, can it then close its stderr pipe? Then from the tool, we could call:

```dart
  final String event = await process.stderr.transform(utf8.decoder).join();
  final Map<String, Object?> result = json.decode(event);
```

If it's not feasible to close stderr after sending the JSON from the other side, then we should catch json decoder errors and in that case, assume we only have part of the message, and store what we have in a variable in the higher scope, and then concatenate the next time the `.listen()` callback fires.

Comment (bkonyi) [reply packages/flutter_tools/lib/src/base/dds.dart:65]: Ah, good point. We've been using similar logic in the standalone VM for years and haven't encountered any reported issues, but it's better to be safe than sorry.

I think we can achieve this using the `JsonDecoder` as a transformer, which will handle the chunking automatically.

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (bkonyi) [reply packages/flutter_tools/lib/src/base/dds.dart:50]: See https://github.com/flutter/flutter/issues/150044

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (bkonyi) [reply packages/flutter_tools/lib/src/base/dds.dart:44]: Done.

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (bkonyi) [reply packages/flutter_tools/lib/src/base/dds.dart]: Ah, `json.decoder` waits for the underlying stream to close before processing the event... closing `stderr` on the DDS side needs to be done as well.

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/dds.dart:53]: Thanks

File: packages/flutter_tools/lib/src/base/dds.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/dds.dart]: Nice, thanks!

File: packages/flutter_tools/lib/src/devtools_launcher.dart
Comment (kenzieschmoll) [inline packages/flutter_tools/lib/src/devtools_launcher.dart:60]: If this wasn't behind the flag would it always print? We don't want the DTD URI to print to CLI in the default case, since this will not be useful for most users.

**Review by kenzieschmoll — APPROVED** [review]

**Comment by kenzieschmoll** [comment]
LGTM from my side, but should wait for @christopherfujino's approval before landing.

File: packages/flutter_tools/lib/src/devtools_launcher.dart
Comment (bkonyi) [reply packages/flutter_tools/lib/src/devtools_launcher.dart:60]: I honestly can't remember what this comment was here for. I'll just remove it.

**Comment by bkonyi** [comment]
@christopherfujino friendly bump.

**Review by christopherfujino — APPROVED** [review]
Code LGTM, but note it looks like you broke g3: cl/646556031/analysis

**Comment by bkonyi** [comment]
The remaining two checks have completed successfully on LUCI even though GitHub doesn't reflect this, so I'm just going to land this.

**Comment by zanderso** [comment]
Reason for revert: Consistently failing `Windows_android native_assets_android` as in https://ci.chromium.org/ui/p/flutter/builders/prod/Windows_android%20native_assets_android/2533/overview

**Pull request merged**
